File: core/src/main/java/dev/morphia/mapping/codec/ReadWritePropertyCodecProvider.java
package dev.morphia.mapping.codec;

import dev.morphia.mapping.codec.kotlin.ReadWritePropertyCodec;
import dev.morphia.mapping.codec.pojo.TypeData;
import kotlin.properties.ReadWriteProperty;
import org.bson.codecs.Codec;
import org.bson.codecs.configuration.CodecConfigurationException;
import org.bson.codecs.pojo.PropertyCodecRegistry;
import org.bson.codecs.pojo.TypeWithTypeParameters;

import java.util.Collection;
import java.util.List;

/**
 * A provider for collection types
 */
@SuppressWarnings("unchecked")
public class ReadWritePropertyCodecProvider extends MorphiaPropertyCodecProvider {
    @Override
    public <T> Codec<T> get(TypeWithTypeParameters<T> type, PropertyCodecRegistry registry) {
        if (ReadWriteProperty.class.isAssignableFrom(type.getType())) {
            final List<? extends TypeWithTypeParameters<?>> typeParameters = type.getTypeParameters();
            TypeWithTypeParameters<?> valueType = getType(typeParameters, 0);

            try {
                return new ReadWritePropertyCodec(registry.get(valueType), type.getType());
            } catch (CodecConfigurationException e) {
                if (valueType.getType().equals(Object.class)) {
                    try {
                        return (Codec<T>) registry.get(TypeData.builder(Collection.class).build());
                    } catch (CodecConfigurationException e1) {
                        // Ignore and return original exception
                    }
                }
                throw e;
            }
        }

        return null;
    }

}

File: kotlin/src/main/kotlin/dev/morphia/mapping/codec/ReadWritePropertyCodecProvider.kt
package dev.morphia.mapping.codec

import com.mongodb.lang.Nullable
import dev.morphia.mapping.codec.pojo.TypeData
import org.bson.codecs.Codec
import org.bson.codecs.configuration.CodecConfigurationException
import org.bson.codecs.pojo.PropertyCodecRegistry
import org.bson.codecs.pojo.TypeWithTypeParameters
import kotlin.properties.ReadWriteProperty

/**
 * @morphia.internal
 * @since 2.2
 */
@Suppress("UNCHECKED_CAST")
class ReadWritePropertyCodecProvider : MorphiaPropertyCodecProvider() {
    @Nullable
    override fun <T> get(type: TypeWithTypeParameters<T>, registry: PropertyCodecRegistry): Codec<T>? {
        if (ReadWriteProperty::class.java.isAssignableFrom(type.type)) {
            val typeParameters = type.typeParameters
            val valueType = getType(typeParameters, 0)
            return try {
                val codec: Codec<Any> = registry[valueType] as Codec<Any>
                ReadWritePropertyCodec<Any>(codec) as Codec<T>
            } catch (e: CodecConfigurationException) {
                if (valueType.type == Any::class.java) {
                    try {
                        return registry.get(TypeData.builder(Collection::class.java).build()) as Codec<T>
                    } catch (_: CodecConfigurationException) {
                        // Ignore and return original exception
                    }
                }
                throw e
            }
        }
        return null
    }
}
